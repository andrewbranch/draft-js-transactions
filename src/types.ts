import { OrderedSet } from 'immutable';

/** Describes the attributes of the characters immediately before and after an offset in a ContentBlock. */
export type NeighboringCharacterAttributes<T> = { before: T | undefined, after: T | undefined };

/**
 * Gets attributes for inserted text based on attributes of characters before
 * and after the insertion position.
 * @param characters The attributes that the neighboring characters at the
 * insertion position will have once preceding edits have been applied.
 */
export type InsertionCallback<T> = (characters: NeighboringCharacterAttributes<T>) => T;

/**
 * Describes where the insertion should occur relative to a selection
 * edge in the case that the insertion takes place at the same offset
 * as an edge of the selection.
 */
export enum SelectionEdgeHandling {
  /** Inserts the text before the selection edge. */
  InsertBefore = 1 << 0,
  /** Inserts the text after the selection edge. */
  InsertAfter = 1 << 1,
  /**
   * Inserts the text inside a non-collapsed selection; that is, after
   * a leading selection edge and before a trailing selection edge.
   * Falls back to `InsertBefore` if the selection is not collapsed.
   */
  InsertInside = 1 << 2,
  /**
   * Inserts the text outside a non-collapsed selection; that is, before
   * a leading selection edge and after a trailing selection edge. Falls
   * back to `InsertBefore` if the selection is not collapsed.
   */
  InsertOutside = 1 << 3
};

/** Describes text inserted by a SliceEdit. */
export type Insertion = {
  /** The characters to insert. */
  text: string;
  /** The inline styles to apply to the inserted text. */
  style?: OrderedSet<string> | InsertionCallback<OrderedSet<string>>;
  /** The key of the Draft Entity to apply to the inserted text. */
  entityKey?: string | null | InsertionCallback<string | null>;
  /**
   * Defines where the insertion should occur relative to a selection edge
   * in the case that the insertion takes place at the same offset as an
   * edge of the selection.
   */
  selectionEdgeHandling?: SelectionEdgeHandling;
};

/**
 * An edit within a block that can delete and insert characters at a specific position.
 */
export type SliceEdit = {
  type: 'slice',
  /** The key of the Draft ContentBlock within which the edit should be performed. */
  blockKey: string;
  /** The character position within the block where the edit should start. */
  offset: number;
  /** The number of characters to remove, starting at `offset`. */
  deletionLength?: number;
  /** The description of the text to insert at `offset`. */
  insertion?: Insertion;
  /**
   * Defines an order of precedence that will be used to decide the order in which
   * edits are applied in the case that multiple edits have the same `offset`.
   * Edits with higher precedence get applied first. If edits have the same offset
   * and the same precedence, they are applied in the order they were added to the
   * transaction.
   */
  precedence?: number;
};
