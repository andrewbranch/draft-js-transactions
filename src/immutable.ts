import { List, Map } from 'immutable';
import { Transaction, Edit } from './types';
import { addEdit, removeEdit, apply } from './transactionBase';

function _createTransaction(edits: Map<string, List<Edit>>): Transaction {
  return {
    addEdit: edit => _createTransaction(addEdit(edits, edit)),
    removeEdit: edit => _createTransaction(removeEdit(edits, edit)),
    apply: editorState => apply(edits, editorState)
  };
}

export function createTransaction(): Transaction {
  const edits: Map<string, List<Edit>> = Map();
  return _createTransaction(edits);
}
