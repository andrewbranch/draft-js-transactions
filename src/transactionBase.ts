import { EditorState, Modifier, SelectionState } from 'draft-js';
import { Collection, List, Map } from 'immutable';
import { Transaction, Edit, NeighboringCharacterAttributes, InsertionCallback, ChangeType } from './types';
import { binaryFindIndex, ComparerFunction } from './utils/binaryFindIndex';
import { identity } from './utils/identity';

const compareEdits: ComparerFunction<Edit> = (a, b) => {
  const offsetOrder = a.offset - b.offset;
  if (offsetOrder !== 0) return offsetOrder;
  return (b.precedence || 0) - (a.precedence || 0);
};

export function addEdit(edits: Map<string, List<Edit>>, edit: Edit): Map<string, List<Edit>> {
  const blockEdits = edits.get(edit.blockKey);
  if (blockEdits) {
    const index = binaryFindIndex(
      blockEdits,
      edit,
      identity,
      compareEdits
    );

    return edits.set(edit.blockKey, blockEdits.splice(index, 0, edit) as List<Edit>) as Map<string, List<Edit>>;
  }

  return edits.set(edit.blockKey, List.of(edit)) as Map<string, List<Edit>>;
};

const isInsertionCallback = <T>(fn: any): fn is InsertionCallback<T> => {
  return typeof fn === 'function';
};

function getAttributes<T extends string | object>(
  attr: T | InsertionCallback<T> | undefined,
  neighboringCharacters: NeighboringCharacterAttributes<T>
): T | undefined {
  if (isInsertionCallback<T>(attr)) {
    return attr(neighboringCharacters);
  }

  return attr;
}

export function apply(edits: Map<string, List<Edit>>, editorState: EditorState, changeType?: ChangeType): EditorState {
  const content = editorState.getCurrentContent();
  const selectionState = editorState.getSelection();
  const anchorOffset = selectionState.getAnchorOffset();
  const focusOffset = selectionState.getFocusOffset();
  const result = edits.reduce((updates, blockEdits) => {
    return blockEdits!.reduce((updates, edit) => {
      const { type, insertion, deletionLength, offset, blockKey } = edit!;
      const shouldMoveAnchor = selectionState.getAnchorKey() === blockKey;
      const shouldMoveFocus = selectionState.getFocusKey() === blockKey;
      switch (type) {
        case 'slice':
          const text = insertion ? insertion.text : '';
          const netInsertionLength = text.length - (deletionLength || 0);
          const position = updates!.offset;
          const block = content.getBlockForKey(blockKey);
          const blockLength = block.getLength();
          const prevCharIndex = position - 1;
          const nextCharIndex = position + (deletionLength || 0);
          const style = getAttributes(insertion ? insertion.style : undefined, {
            before: prevCharIndex < 0 ? undefined : block.getInlineStyleAt(prevCharIndex),
            after: nextCharIndex > blockLength - 1 ? undefined : block.getInlineStyleAt(nextCharIndex)
          });
          const entityKey = getAttributes(insertion ? insertion.entityKey : undefined, {
            before: prevCharIndex < 0 ? undefined : block.getEntityAt(prevCharIndex),
            after: nextCharIndex > blockLength - 1 ? undefined : block.getEntityAt(nextCharIndex)
          });
          const anchorOffset = Math.max(updates!.deletionEnd, offset);
          const focusOffset = Math.max(anchorOffset + (deletionLength || 0) - (anchorOffset - offset), anchorOffset);
          const selectionToReplace = SelectionState.createEmpty(blockKey).merge({
            anchorOffset: updates!.offset + anchorOffset,
            focusOffset: updates!.offset + focusOffset
          }) as SelectionState;

          return {
            content: Modifier.replaceText(updates!.content, selectionToReplace, text, style, entityKey),
            offset: updates!.offset + netInsertionLength,
            deletionEnd: updates!.offset + focusOffset,
            changeType: insertion
              ? 'insert-characters' : deletionLength
              ? updates!.changeType || 'remove-range'
              : null
          };
        default:
          ((t: never) => {
            throw new Error(`Unrecognized edit type: ${type}`);
          })(type);
      }
    }, updates);
  }, {
    content,
    offset: 0,
    deletionEnd: 0,
    changeType: null as ChangeType | null
  });

  return EditorState.push(
    editorState,
    result.content,
    changeType || result.changeType || 'insert-characters'
  );
};
