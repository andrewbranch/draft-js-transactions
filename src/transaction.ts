import { List, Map } from 'immutable';
import { EditorState } from 'draft-js';
import { Transaction as ITransaction, Edit } from './types';
import { addEdit, apply } from './transactionBase';

export class Transaction implements ITransaction {
  private edits: Map<string, List<Edit>> = Map();

  public get size() {
    return this.edits.size;
  }

  public addEdit(edit: Edit) {
    this.edits = addEdit(this.edits, edit);
    return this;
  }
  
  public apply(editorState: EditorState) {
    return apply(this.edits, editorState);
  }
}
