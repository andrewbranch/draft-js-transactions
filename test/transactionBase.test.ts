import { EditorState, ContentState, SelectionState } from 'draft-js';
import { Map, List, OrderedSet } from 'immutable';
import { Edit } from '../src/types';
import { addEdit, apply } from '../src/transactionBase';

const createEditorState = (text?: string) => EditorState.createWithContent(ContentState.createFromText(text || ''));

describe('transactionBase', () => {
  describe('addEdit', () => {
    test('inserts edits by block', () => {
      let editMap: Map<string, List<Edit>> = Map();
      const edits: Edit[] = [
        { type: 'splice', blockKey: '1', offset: 0 },
        { type: 'splice', blockKey: '1', offset: 0 },
        { type: 'splice', blockKey: '2', offset: 0 }
      ];

      editMap = addEdit(editMap, edits[0]);
      editMap = addEdit(editMap, edits[1]);
      editMap = addEdit(editMap, edits[2]);
      expect(editMap).toEqual(Map({
        1: List.of(edits[0], edits[1]),
        2: List.of(edits[2])
      }));
    });

    test('inserts edits ordered by offset', () => {
      let editMap: Map<string, List<Edit>> = Map();
      const edits: Edit[] = [
        { type: 'splice', blockKey: '1', offset: 2 },
        { type: 'splice', blockKey: '1', offset: 0 },
        { type: 'splice', blockKey: '1', offset: 1 }
      ];

      editMap = addEdit(editMap, edits[0]);
      editMap = addEdit(editMap, edits[1]);
      editMap = addEdit(editMap, edits[2]);
      expect(editMap).toEqual(Map({
        1: List.of(edits[1], edits[2], edits[0])
      }));
    });

    test('inserts edits of equal offset ordered by precedence', () => {
      let editMap: Map<string, List<Edit>> = Map();
      const edits: Edit[] = [
        { type: 'splice', blockKey: '1', offset: 0, precedence: 2 },
        { type: 'splice', blockKey: '1', offset: 0, precedence: 0 },
        { type: 'splice', blockKey: '1', offset: 0, precedence: 1 }
      ];

      editMap = addEdit(editMap, edits[0]);
      editMap = addEdit(editMap, edits[1]);
      editMap = addEdit(editMap, edits[2]);
      expect(editMap).toEqual(Map({
        1: List.of(edits[0], edits[2], edits[1])
      }));
    });
  });

 
  describe('apply', () => {
    describe('error handling', () => {
      test('unrecognized edit type throws', () => {
        expect(() => {
          apply(addEdit(Map(), { type: 'asdfjkl;' as any, blockKey: '1', offset: 0 }), createEditorState());
        }).toThrow(/unrecognized edit type/i);
      });

      test('unrecognized selection edge handling throws', () => {
        expect(() => {
          const editorState = createEditorState();
          apply(addEdit(Map(), {
            type: 'splice',
            blockKey: editorState.getCurrentContent().getFirstBlock().getKey(),
            offset: 0,
            insertion: { text: '', selectionEdgeHandling: 'asdfjk;' as any }
          }), editorState);
        }).toThrow(/unrecognized selection ?edge ?handling/i);
      });
    });

    describe('basic insertions and deletions', () => {
      test('can insert characters at multiple positions in a block', () => {
        const editorState = createEditorState('one two');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 0,
          insertion: { text: 'zero ' }
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          insertion: { text: 'one-and-a-half ' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('zero one one-and-a-half two');
      });

      test('can delete characters at multiple positions in a block', () => {
        const editorState = createEditorState('one two three');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 0,
          deletionLength: 4
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 7,
          deletionLength: 6
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('two');
      });

      test('can delete and insert with a single edit', () => {
        const editorState = createEditorState('one two three');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 3,
          insertion: { text: 'dos' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one dos three');
      });

      test('overlapping deletions delete the union of the deletion range', () => {
        const editorState = createEditorState('a b c d e f');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 0,
          deletionLength: 4,
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 2,
          deletionLength: 4
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('d e f');
      });

      test('can replace multiple ranges with longer insertions than deletions', () => {
        const editorState = createEditorState('one x three x five');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 1,
          insertion: { text: 'two' }
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 12,
          deletionLength: 1,
          insertion: { text: 'four' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one two three four five');
      });

      test('can replace multiple ranges with shorter insertions than deletions', () => {
        const editorState = createEditorState('one two three four five');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 3,
          insertion: { text: 'x' }
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 14,
          deletionLength: 4,
          insertion: { text: 'x' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one x three x five');
      });

      test('can insert text into a deleted range', () => {
        const editorState = createEditorState('one two three');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 3,
          insertion: { text: '' }
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 0,
          insertion: { text: 'x' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one x three');
      });

      test('can replace text in a deleted range', () => {
        const editorState = createEditorState('one two three');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 3,
          insertion: { text: '' }
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 2,
          insertion: { text: 'x' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one x three');
      });


      test('can replace text overlapping deleted range', () => {
        const editorState = createEditorState('one two three');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 3,
          insertion: { text: '' }
        });
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[0].getKey(),
          offset: 4,
          deletionLength: 4,
          insertion: { text: 'x ' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one x three');
      });

      test('edits occur in the correct block', () => {
        const editorState = createEditorState('one two\nthree');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[1].getKey(),
          offset: 0,
          deletionLength: 5,
          insertion: { text: 'four' }
        });

        expect(apply(editMap, editorState).getCurrentContent().getPlainText()).toBe('one two\nfour');
      });
    });

    describe('styles and entities', () => {
      test('can insert text with inline styles', () => {
        const editorState = createEditorState();
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 0,
          insertion: {
            text: 'Bold',
            style: OrderedSet.of('BOLD')
          }
        });

        expect(apply(editMap, editorState).getCurrentContent().getFirstBlock().getCharacterList()).toMatchSnapshot();
      });

      test('can insert text with an entity key', () => {
        const editorState = createEditorState();
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 0,
          insertion: {
            text: 'Entity',
            entityKey: '1'
          }
        });

        expect(apply(editMap, editorState).getCurrentContent().getFirstBlock().getCharacterList()).toMatchSnapshot();
      });

      test('can insert text with a style callback', () => {
        let editorState = createEditorState();
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        let setupEdits: Map<string, List<Edit>> = Map();
        setupEdits = addEdit(setupEdits, {
          type: 'splice',
          blockKey,
          offset: 0,
          insertion: {
            text: 'bold',
            style: OrderedSet.of('BOLD')
          }
        });
        setupEdits = addEdit(setupEdits, {
          type: 'splice',
          blockKey,
          offset: 4,
          insertion: {
            text: 'italic',
            style: OrderedSet.of('ITALIC')
          }
        });
        editorState = apply(setupEdits, editorState);

        let testEdits: Map<string, List<Edit>> = Map();
        testEdits = addEdit(testEdits, {
          type: 'splice',
          blockKey,
          offset: 4,
          insertion: {
            text: 'both',
            style: ({ before, after }) => {
              expect(before).toEqual(OrderedSet.of('BOLD'));
              expect(after).toEqual(OrderedSet.of('ITALIC'));
              return before!.union(after!);
            }
          }
        });

        expect(apply(testEdits, editorState).getCurrentContent().getFirstBlock().getCharacterList()).toMatchSnapshot();
      });
    });

    describe('selection behavior', () => {
      test('inserting text before a selection moves selection forward', () => {
        let editorState = createEditorState('one two');
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        editorState = EditorState.forceSelection(editorState, editorState.getSelection().merge({
          anchorOffset: 4,
          focusOffset: 4
        }) as SelectionState);

        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 0,
          insertion: { text: 'zero' }
        });

        expect(apply(editMap, editorState).getSelection().toJS()).toMatchObject({
          anchorKey: blockKey,
          anchorOffset: 8,
          focusKey: blockKey,
          focusOffset: 8,
          hasFocus: true,
          isBackward: false
        });
      });

      test('deleting text before a selection moves selection back', () => {
        let editorState = createEditorState('one two');
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        editorState = EditorState.forceSelection(editorState, editorState.getSelection().merge({
          anchorOffset: 4,
          focusOffset: 4
        }) as SelectionState);

        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 0,
          deletionLength: 4
        });

        expect(apply(editMap, editorState).getSelection().toJS()).toMatchObject({
          anchorKey: blockKey,
          anchorOffset: 0,
          focusKey: blockKey,
          focusOffset: 0,
          hasFocus: true,
          isBackward: false
        });
      });

      test('inserting text after a selection doesn’t move the selection', () => {
        let editorState = createEditorState('one two');
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();

        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 7,
          insertion: { text: ' three' }
        });

        expect(apply(editMap, editorState).getSelection().toJS()).toMatchObject({
          anchorKey: blockKey,
          anchorOffset: 0,
          focusKey: blockKey,
          focusOffset: 0,
          hasFocus: true,
          isBackward: false
        });
      });

      test('inserting text in another block doesn’t move the selection', () => {
        let editorState = createEditorState('one two\nthree');
        const blocks = editorState.getCurrentContent().getBlocksAsArray();
        editorState = EditorState.forceSelection(editorState, editorState.getSelection().merge({
          anchorOffset: 4,
          focusOffset: 4
        }) as SelectionState);

        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey: blocks[1].getKey(),
          offset: 0,
          insertion: { text: 'four ' }
        });

        expect(apply(editMap, editorState).getSelection().toJS()).toMatchObject({
          anchorKey: blocks[0].getKey(),
          anchorOffset: 4,
          focusKey: blocks[0].getKey(),
          focusOffset: 4,
          hasFocus: true,
          isBackward: false
        });
      });

      test('replacing text before a selection moves the selection by the net insertion length', () => {
        let editorState = createEditorState('one two');
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        editorState = EditorState.forceSelection(editorState, editorState.getSelection().merge({
          anchorOffset: 4,
          focusOffset: 4
        }) as SelectionState);

        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 0,
          deletionLength: 3,
          insertion: { text: 'four' }
        });

        expect(apply(editMap, editorState).getSelection().toJS()).toMatchObject({
          anchorKey: blockKey,
          anchorOffset: 5,
          focusKey: blockKey,
          focusOffset: 5,
          hasFocus: true,
          isBackward: false
        });
      });

      test('deleting text around a selection moves the selection to the beginning of the deleted range', () => {
        let editorState = createEditorState('one two');
        const blockKey = editorState.getCurrentContent().getFirstBlock().getKey();
        editorState = EditorState.forceSelection(editorState, editorState.getSelection().merge({
          anchorOffset: 2,
          focusOffset: 2
        }) as SelectionState);

        let editMap: Map<string, List<Edit>> = Map();
        editMap = addEdit(editMap, {
          type: 'splice',
          blockKey,
          offset: 1,
          deletionLength: 2
        });

        expect(apply(editMap, editorState).getSelection().toJS()).toMatchObject({
          anchorKey: blockKey,
          anchorOffset: 1,
          focusKey: blockKey,
          focusOffset: 1,
          hasFocus: true,
          isBackward: false
        });
      });
    });
  });
});
