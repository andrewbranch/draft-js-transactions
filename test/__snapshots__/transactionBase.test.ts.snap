// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactionBase apply styles and entities can insert text with a style callback 1`] = `
Immutable.List [
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "ITALIC",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "ITALIC",
    ],
    "entity": null,
  },
]
`;

exports[`transactionBase apply styles and entities can insert text with an entity key 1`] = `
Immutable.List [
  Immutable.Record {
    "style": Immutable.OrderedSet [],
    "entity": "1",
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [],
    "entity": "1",
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [],
    "entity": "1",
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [],
    "entity": "1",
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [],
    "entity": "1",
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [],
    "entity": "1",
  },
]
`;

exports[`transactionBase apply styles and entities can insert text with inline styles 1`] = `
Immutable.List [
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
  Immutable.Record {
    "style": Immutable.OrderedSet [
      "BOLD",
    ],
    "entity": null,
  },
]
`;
